### PR TITLE
Limit the types for kubectl expose command

### PR DESCRIPTION
**Pull request by feihujiang**
Fixes the first part of #13178

### CODE REVIEW

File: pkg/kubectl/cmd/expose.go [hunk 1]
```diff
@@ -106,6 +106,9 @@ func RunExpose(f *cmdutil.Factory, out io.Writer, cmd *cobra.Command, args []str
 		return fmt.Errorf("multiple resources provided: %v", args)
 	}
 	info := infos[0]
+	if info.Mapping.Kind != "ReplicationController" && info.Mapping.Kind != "Service" {
+		return fmt.Errorf("kubectl expose must use a replication controller or service, not use a %v", info.Mapping.Kind)
+	}
```
Comment (hurf) [inline pkg/kubectl/cmd/expose.go]: What about `invalid resouce type: %v, only rc and service are accepted.`?

File: pkg/kubectl/cmd/expose.go [hunk 2]
```diff
@@ -106,6 +106,9 @@ func RunExpose(f *cmdutil.Factory, out io.Writer, cmd *cobra.Command, args []str
 		return fmt.Errorf("multiple resources provided: %v", args)
 	}
 	info := infos[0]
+	if info.Mapping.Kind != "ReplicationController" && info.Mapping.Kind != "Service" {
+		return fmt.Errorf("kubectl expose must use a replication controller or service, not use a %v", info.Mapping.Kind)
```
Comment (hurf) [inline pkg/kubectl/cmd/expose.go]: What about `invalid resource prodived: %v`

File: pkg/kubectl/cmd/expose.go [hunk 3]
```diff
@@ -106,6 +109,9 @@ func RunExpose(f *cmdutil.Factory, out io.Writer, cmd *cobra.Command, args []str
 		return fmt.Errorf("multiple resources provided: %v", args)
 	}
 	info := infos[0]
+	if info.Mapping.Kind != "ReplicationController" && info.Mapping.Kind != "Service" && info.Mapping.Kind != "Pod" {
+		return fmt.Errorf("invalid resource prodived: %v, only a replication controller, service or pod are accepted", info.Mapping.Kind)
```
Comment (JanetKuo) [inline pkg/kubectl/cmd/expose.go]: change "are" to "is"

File: pkg/kubectl/cmd/expose.go [hunk 4]
```diff
@@ -30,7 +30,7 @@ import (
 const (
 	expose_long = `Take a replicated application and expose it as Kubernetes Service.
```
Comment (JanetKuo) [inline pkg/kubectl/cmd/expose.go]: include "pod" and "service" here?

**Comment by JanetKuo** [comment]
Fix the nits then LGTM.

**Comment by feihujiang** [comment]
@JanetKuo Changed according to your advcie,   please review.

**Comment by JanetKuo** [comment]
We need rebase now.

**Comment by JanetKuo** [comment]
Hi @feihujiang please rebase again and ping me when you're done.

File: pkg/kubectl/cmd/expose.go [hunk 5]
```diff
@@ -112,6 +115,9 @@ func RunExpose(f *cmdutil.Factory, out io.Writer, cmd *cobra.Command, args []str
 		return fmt.Errorf("multiple resources provided: %v", args)
 	}
 	info := infos[0]
+	if info.Mapping.Kind != "ReplicationController" && info.Mapping.Kind != "Service" && info.Mapping.Kind != "Pod" {
+		return fmt.Errorf("invalid resource prodived: %v, only a replication controller, service or pod is accepted", info.Mapping.Kind)
```
Comment (ArtfulCoder) [inline pkg/kubectl/cmd/expose.go]: typo in spelling "prodived"

**Comment by feihujiang** [comment]
@JanetKuo Fixed the nit and rebased, please review.

File: pkg/kubectl/cmd/expose.go [hunk 6]
```diff
@@ -112,6 +115,9 @@ func RunExpose(f *cmdutil.Factory, out io.Writer, cmd *cobra.Command, args []str
 		return fmt.Errorf("multiple resources provided: %v", args)
 	}
 	info := infos[0]
+	if info.Mapping.Kind != "ReplicationController" && info.Mapping.Kind != "Service" && info.Mapping.Kind != "Pod" {
```
Comment (kargakis) [inline pkg/kubectl/cmd/expose.go]: Do we want this? At the very least, we could/should achieve this through the factory

cc: @smarterclayton

**Comment by smarterclayton** [comment]
@kargakis review this for our use of the expose command for routes.

**Comment by smarterclayton** [comment]
Please do not merge until kargakis has a chance to review

File: pkg/kubectl/cmd/expose.go (hunk 6 above)
Comment (kargakis) [inline pkg/kubectl/cmd/expose.go]: @feihujiang create a `CanBeExposed func(kind string) bool` method in the factory and handle this check in [there](https://github.com/kubernetes/kubernetes/blob/196f58b9cb25a2222c7f9aacd624737910b03acb/pkg/kubectl/cmd/util/factory.go#L50).

**Comment by feihujiang** [comment]
@kargakis  Updated based on your feedback, please review.

File: pkg/kubectl/cmd/util/factory.go [hunk 7]
```diff
@@ -83,6 +83,8 @@ type Factory struct {
 	DefaultNamespace func() (string, bool, error)
 	// Returns the generator for the provided generator name
 	Generator func(name string) (kubectl.Generator, bool)
+	// Check whether the kind of resources could be exposed
+	CanBeExposed func(kind string) error
```
Comment (kargakis) [inline pkg/kubectl/cmd/util/factory.go]: @smarterclayton any strong preference between `func(kind string)` and `func(obj runtime.Object)`?

Comment (smarterclayton) [inline pkg/kubectl/cmd/util/factory.go]: kind is probably better

On Wed, Sep 9, 2015 at 3:09 AM, Michail Kargakis <notifications@github.com>
wrote:

> In pkg/kubectl/cmd/util/factory.go
> <https://github.com/kubernetes/kubernetes/pull/13184#discussion_r39012619>
> :
>
> > @@ -83,6 +83,8 @@ type Factory struct {
> >  	DefaultNamespace func() (string, bool, error)
> >  	// Returns the generator for the provided generator name
> >  	Generator func(name string) (kubectl.Generator, bool)
> > +	// Check whether the kind of resources could be exposed
> > +	CanBeExposed func(kind string) error
>
> @smarterclayton <https://github.com/smarterclayton> any strong preference
> between func(kind string) and func(obj runtime.Object)?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/13184/files#r39012619>.
>



-- 
Clayton Coleman | Lead Engineer, OpenShift

**Comment by kargakis** [comment]
@feihujiang can you add a negative test in `hack/test-cmd.sh`?

**Comment by feihujiang** [comment]
@kargakis I try to add a negative test,  command:"./cluster/kubectl.sh  expose nodes 127.0.0.1 --port=80", postcondition:no service exposed. but if kubectl doesn't have this commit , service also can't be exposed, an error message "error: couldn't find selectors via --selector flag or introspection: cannot extract pod selector from Node". Could you give me an advice to add a negative test about this commit?

**Comment by kargakis** [comment]
You have to hit the "invalid resource provided" error returned from `CanBeExposed`. I am surprised you don't get it when trying to expose a node.

**Comment by feihujiang** [comment]
@kargakis I added tests (negative and positive) for CanBeExposed, please review.

**Comment by kargakis** [comment]
Unit tests LGTM but this also needs to be tested in hack/test-cmd. Try to grep the error message returned from kubectl expose node

Something like 
```bash
 [[ "$(kubectl expose node 127.0.0.1 "${kube_flags[@]}" | grep 'invalid resource')" ]]
```

**Comment by kargakis** [comment]
@feihujiang I see you added the test I asked and everything is green! Thanks!

LGTM

**Comment by feihujiang** [comment]
@kargakis  Thanks for your review.

**Comment by smarterclayton** [comment]
Needs rebase, then I'll add the merge tag

**Comment by feihujiang** [comment]
@smarterclayton I rebased, please review.

**Comment by feihujiang** [comment]
@k8s-bot test this please

**Comment by feihujiang** [comment]
@JanetKuo I rebased,  please review. And could you please merge this commit earlier, thanks!

**Comment by feihujiang** [comment]
@smarterclayton I rebased, please give this a lgtm, thanks!

**Comment by smarterclayton** [comment]
Shippable looked like a flake, rerunning

**Comment by feihujiang** [comment]
@k8s-bot test this please

**Comment by feihujiang** [comment]
@JanetKuo I have rebased this commit more than three times, Could you please merge this commit as soon as possible, thanks!

**Comment by JanetKuo** [comment]
@feihujiang only @k8s-merge-robot or oncalls can merge PRs. I've added LGTM label again. It should be merged soon.

**Comment by feihujiang** [comment]
@JanetKuo I got it, thanks very much! It's a little boring to rebase many times. (×_×)

**Pull request merged**